### PR TITLE
Feature remote control highlight

### PR DESCRIPTION
**Pull request by singhsanket143**
The pull request is implementation for the feature listed in issue #16863. Implementation of Remote control of a projected instance from another device has been implemented in this pull request. The package named yuukan:streamy has been used to establish real-time communication between different devices. Whenever the admin checks the highlight instance checkbox all the other devices will show the effect of emphasizing and de-emphasizing on them.

1. From the admin end.
image

2. From the other user end
image

### CODE REVIEW

**Review by peter-hank — CHANGES_REQUESTED** [review]
@singhsanket143 the whole implementation is on the client side and people with skills will be able to fake it, can you make sure it's not possible?

File: server/methods.js [hunk 1]
```diff
@@ -759,5 +759,35 @@ Meteor.methods({
...
+    if (!this.userId) {
+      return false;
+    }
+    if (adminId) {
+      user = Meteor.users.findOne({ _id: adminId });
+    } else {
+      return false;
+    }
+    console.log(user.emails[0].address, table.admin );
+    return (user.emails[0].address === table.admin );
```
Comment (peter-hank) [inline server/methods.js]: @singhsanket143 can you remote the extra space before `);`?

File: server/methods.js [hunk 2]
```diff
@@ -759,5 +759,35 @@ Meteor.methods({
...
+    if (table === undefined) { return false; }
+    if (!this.userId) {
+      return false;
+    }
+    if (adminId) {
+      user = Meteor.users.findOne({ _id: adminId });
+    } else {
+      return false;
+    }
+    console.log(user.emails[0].address, table.admin );
```
Comment (peter-hank) [inline server/methods.js]: @singhsanket143 can you remove it?

File: server/methods.js [hunk 3]
```diff
@@ -759,5 +759,35 @@ Meteor.methods({
     }
     console.log("OK");
     return true;
+  },
+  instanceAdminCheck(instanceid, adminId) {
```
Comment (peter-hank) [inline server/methods.js]: @singhsanket143 you don't need to (and shouldn't) pass the `adminId` here. Instead, us `Meteor.userId()` to get the current user id.

File: server/methods.js [hunk 4]
```diff
@@ -759,5 +759,35 @@ Meteor.methods({
...
+    }
+    if (adminId) {
+      user = Meteor.users.findOne({ _id: adminId });
+    } else {
+      return false;
+    }
+    console.log(user.emails[0].address, table.admin );
+    return (user.emails[0].address === table.admin );
+  },
+  emphasize(divId, instanceId, currUser) {
```
Comment (peter-hank) [inline server/methods.js]: @singhsanket143 https://github.com/berkmancenter/question_tool/pull/134/files#r316066171

the same here

File: server/methods.js [hunk 5]
```diff
@@ -759,5 +759,35 @@ Meteor.methods({
...
+    let currInstance = Instances.findOne({ _id: instanceId });
+    if (currInstance.admin === currUser.emails[0].address) {
+      let adminOfInstance = Meteor.users.findOne({'services.password.reset.email': currInstance.admin});
+      Streamy.broadcast('emphasize', { id: divId, instanceId: instanceId, admin: adminOfInstance._id });
+    }
+  },
+  deEmphasize(divId, instanceId, currUser) {
```
Comment (peter-hank) [inline server/methods.js]: @singhsanket143 https://github.com/berkmancenter/question_tool/pull/134/files#r316066171

the same here

File: client/views/helpers/chunks/question_div/question_div.js [hunk 6]
```diff
@@ -53,6 +71,10 @@ Template.question_div.helpers({
   time_format(timeorder) {
     return moment(timeorder).fromNow();
   },
+
+  isAdmin() {
```
Comment (peter-hank) [inline client/views/helpers/chunks/question_div/question_div.js]: @singhsanket143 can you move this check to the server part as well?

**Review by peter-hank — CHANGES_REQUESTED** [review]
A few more changes.

File: client/views/helpers/chunks/question_div/question_div.js [hunk 7]
```diff
@@ -128,6 +128,7 @@ Template.question_div.events({
     if (currInstance.admin !== Meteor.user().emails[0].address) {
       return;
     }
+    console.log(Meteor.user());
```
Comment (peter-hank) [inline client/views/helpers/chunks/question_div/question_div.js]: @singhsanket143 can you remove it?

**Review by peter-hank — CHANGES_REQUESTED** [review]
@singhsanket143 

Not all of the requested changes are done. Can you go through this again?

**Review by peter-hank — CHANGES_REQUESTED** [review]
It's still overcomplicated.

The logic should be:
- click on the highlight button
- check on the server part if the user is admin
- if so, broadcast and apply on the client side, if not don't do anything

Can you try to simplify it?

**Comment by singhsanket143** [comment]
> It's still overcomplicated.
> 
> The logic should be:
> 
> * click on the highlight button
> * check on the server part if the user is admin
> * if so, broadcast and apply on the client side, if not don't do anything
> 
> Can you try to simplify it?

@peter-hank I have tried to simplify the code. Please review it.

**Comment by peter-hank** [comment]
@singhsanket143 can you make the `Highlight the question` label clickable?

**Comment by singhsanket143** [comment]
> @singhsanket143 can you make the `Highlight the question` label clickable?

@peter-hank So on clicking checkbox as well as on label the highlight action should occur?

**Comment by singhsanket143** [comment]
> @singhsanket143 can you make the `Highlight the question` label clickable?

@peter-hank added the clickable label

File: client/views/helpers/chunks/question_div/question_div.html [hunk 8]
```diff
@@ -62,7 +62,7 @@
       </div>
       {{#if adminButtons}}
         <div class="question-highlight">
-          <input class="remote-highlight" type="checkbox" name="highlight" value="0"> Highlight the question
+          <label class="remote-highlight"><input type="checkbox" name="highlight" value="0">Highlight the question</label>
```
Comment (peter-hank) [inline client/views/helpers/chunks/question_div/question_div.html]: @singhsanket143 now the label is a little bit off. Can you add some space between the label and the checkbox and make the label vertically centered? Also, make the cursor be `pointer` on hover.

**Review by peter-hank — CHANGES_REQUESTED** [review]

File: client/views/helpers/chunks/question_div/question_div.html (hunk 8 above)
Comment (singhsanket143) [reply client/views/helpers/chunks/question_div/question_div.html]: @peter-hank I have made the corresponding changes

File: client/views/helpers/chunks/question_div/question_div.scss [hunk 9]
```diff
@@ -390,3 +390,24 @@ $popular-question-color: #f1c40f;
...
+
+.remote-highlight > input {
+  width: 13px;
+  height: 13px;
+  padding: 0;
+  margin:0;
+  vertical-align: bottom;
+  position: relative;
+  top: -1px;
+  *overflow: hidden;
```
Comment (peter-hank) [inline client/views/helpers/chunks/question_div/question_div.scss]: @singhsanket143 any good reason to have `*` here?

**Review by peter-hank — CHANGES_REQUESTED** [review]

File: client/views/helpers/chunks/question_div/question_div.scss [hunk 10]
```diff
@@ -390,3 +390,24 @@ $popular-question-color: #f1c40f;
   font-weight: 600;
   color: #777!important;
 }
+
+.remote-highlight:hover, .remote-highlight > input:hover {
```
Comment (peter-hank) [inline client/views/helpers/chunks/question_div/question_div.scss]: @singhsanket143 isn't `.remote-highlight:hover` enough here?

File: client/views/helpers/chunks/question_div/question_div.scss (hunk 9 above)
Comment (singhsanket143) [reply client/views/helpers/chunks/question_div/question_div.scss]: @peter-hank I have removed this.

File: client/views/helpers/chunks/question_div/question_div.scss (hunk 10 above)
Comment (singhsanket143) [reply client/views/helpers/chunks/question_div/question_div.scss]: @peter-hank Updated this css styling.

**Review by peter-hank — APPROVED** [review]

**Pull request merged**